R/preprocessing.R

```diff
@@ -1526,3 +1526,165 @@ RegressOutNB <- function(
+        dplyr::group_by(orig.ident) %>%
+        dplyr::mutate(rank = rank(-!!rlang::sym(umi.column))) %>%
+        dplyr::arrange(orig.ident, rank)
+
+    ## Set thresholds for rank of barcodes to ignore
+    if (is.null(threshold.low)) {
+        threshold.low <- 1
+    }
+    if (is.null(threshold.high)) {
+        threshold.high <- max(tmp_meta$rank)
+    }
+
+    ## Calculate inflection point within each group's barcode range
+    inflections <- tmp_meta %>%
+        tidyr::nest() %>%
+        dplyr::mutate(inflection = purrr::map_int(data, function(x) {
+            y <- x %>%
+                dplyr::mutate(
+                    rawdiff = c(NA, diff(log10(!!rlang::sym(umi.column) + 1)) / diff(rank))
+                ) %>%
+                dplyr::summarise(
+                    inflection = which(rawdiff == min(rawdiff[threshold.low:threshold.high], na.rm = TRUE))
+                )
+            y[[1]]
+        })) %>%
+        dplyr::select(-data)
+
+    ## Calculate inflection point plot for QC
+    if (show.plot == TRUE | return.plot == TRUE) {
+        g <- tmp_meta %>%
+            ggplot2::ggplot(aes(x = rank, y = log10(!!rlang::sym(umi.column)),
+                       group = orig.ident, colour = orig.ident)) +
+            ggplot2::geom_line() +
+            ggplot2::theme_bw() +
+            ggplot2::geom_vline(data = inflections,
+                               aes(xintercept = inflection,
+                                   group = orig.ident, colour = orig.ident),
+                               linetype = 'dashed')
+    }
+
+    ## Show the plot if desired
+    if (show.plot == TRUE) {
+        print(g)
+    }
+
+    ## Break: return ggplot2 object
+    if (return.plot == TRUE) {
+        return(g)
+    }
+
+    ## Filter the cells based on calculated inflection points per sample
+    cells_filtered <- tmp_meta %>%
+        dplyr::select(orig.ident, cell, !!rlang::sym(umi.column), rank) %>%
+        dplyr::group_by(orig.ident) %>%
+        tidyr::nest() %>%
+        dplyr::inner_join(inflections, by = 'orig.ident') %>%
+        dplyr::mutate(
+            umi_threshold = purrr::map2_dbl(data, inflection, function(d, i) {
+                x <- dplyr::filter(d, rank < i)
+                min(x[umi.column], na.rm = TRUE) # return the min nCount threshold used
+            }),
+            cells_pass = purrr::map2(data, inflection, function(d, i) {
+                x <- dplyr::filter(d, rank < i)
+                return(x$cell) # return passing cells
+            }))
+
+    ## Break: returns the calculated inflection points data and passing cells
+    if (return.inflections) {
+        return(cells_filtered)
+    }
+    
+    ## Get vector of which cells to keep
+    cells_pass <- unlist(cells_filtered$cells_pass)
+
+    ## Break: returns the names of the cells that pass only
+    if (return.cells == TRUE) {
+        return(cells_pass)
+    }
+
+    ## Subset the object object
+    object <- SubsetData(object, cells = cells_pass)
```

Comment:
`SubsetData` will be deprecated at some point in the near future, could you use `subset` or the `[` operator instead? Please see `?subset.Seruat` for details

Reply:
Do you think that directly subsetting the data and returning the subsetted object is most appropriate for the function? Or, should it be that it returns the `meta.data` slot with an appended column `cells_pass` with a boolean of whether it is above the "knee" or not? 

If so, is there a way to make it so that the new working data is the subsetted version (and thus doesn't lose the "bad cells")?

Probably easiest to just return the subsetted Seurat object that removes the "bad cells", but wanted to check in on this

Reply:
If the function is designed to subset (since it's called `SubsetByBarcodeFraction`, it sounds like a subsetting function to me), then it should return a subsetted object, following how Seurat subsets objects. `SubsetData`, for example, subsets all aspects of the Seurat object down to certain features and/or cells.

If the function was to simply calculate the inflection point, but not subset, then it should be titled differently. I personally think that as long as you're adding something to object meta data that users can subset on themselves, you can leave subsetting up to the user.